include/utils.hpp

```diff
@@ -126,6 +137,28 @@ precondition(bool cond) {
 }
 #endif
 
+template<auto X>
+inline constexpr std::integral_constant<std::remove_const_t<decltype(X)>, X> ic{};
+
+/**
+ * T is tuple-like if it implements std::tuple_size, std::tuple_element, and std::get.
+ * Tuples with size 0 are excluded.
+ */
+template<typename T>
+concept tuple_like = (std::tuple_size<T>::value > 0) && requires(T tup) {
```

Comment:
This is similar to [p2165](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p2165r2.pdf), isn't it?

Reply:
Yes, except that I excluded the size 0 cases.